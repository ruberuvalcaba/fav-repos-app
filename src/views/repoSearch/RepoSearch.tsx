import React, { useState, useEffect, useContext } from 'react'
import { Container, Input, Box, Spinner, Text } from '@chakra-ui/react'
import { SearchIcon, CloseIcon, WarningTwoIcon } from '@chakra-ui/icons'
import SearchRepoResults from './RepoSearchResults'
import { RepoContext } from '../../store/ReposStore'
import API from '../../api'
import useDebounce from '../../hooks/useDebounce'
import { Repo, RepoPayload, AlerBarStatus } from '../../types'
import { AlertBar } from '../sharedComponents'

const SearchRepo = (): JSX.Element => {
  const [searchValue, setSearchValue] = useState<string>('')
  const [reposListResults, setReposListResults] = useState<Repo[]>([])
  const [isLoading, setIsLoading] = useState<boolean>(false)
  const [isAlertVisible, setIsAlertVisible] = useState<boolean>(false)
  const [showNoResultsMessage, setShowNoResultsMessage] = useState<boolean>(false)
  const [isServerRunning, setIsServerRunning] = useState<boolean>(false)
  const [alertInfo, setAlertInfo] = useState<{
    message: string
    description?: string
    status: AlerBarStatus
  }>({ message: '', description: '', status: 'success' })
  const { state, actions } = useContext(RepoContext)
  const { reposList } = state
  const debouncedSearchValue = useDebounce(searchValue, 400)

  /**
   * Helps on preventing interaction with the server when it isn't running.
   */
  const getIsReposerverRunning = async () => {
    const response = await API.healthCheck()
    if (response && response?.status === 200) {
      setIsServerRunning(true)
      return true
    } else {
      setIsServerRunning(false)
      return false
    }
    return
  }

  /*
   * Makes a health check on loading before allowing users to start searching for repos
   */
  useEffect(() => {
    getIsReposerverRunning()
  }, [])

  /*
   * filter results to suggest a more accurate list based on user input
   */
  const filterAndFormatResults = (results: RepoPayload[]): Repo[] => {
    const searchInput = searchValue.trim()
    let regex: RegExp
    try {
      regex = new RegExp(`^${searchInput}`, 'i') //Prevents odd imputs ie. (#$&*#{{F/
    } catch (e) {
      console.error('Invalid search input')
    }
    return results
      .filter((item: RepoPayload) =>
        regex
          ? regex.test(item.name) || regex.test(item.description) || regex.test(item.full_name)
          : item
      )
      .map((item: RepoPayload) => {
        const { id, name, full_name, description, url, created_at, stargazers_count, language } =
          item || {}
        return {
          id: id.toString(),
          name,
          fullName: full_name,
          description,
          url,
          createdAt: created_at,
          stargazersCount: stargazers_count,
          language,
        }
      })
      .sort((a: Repo, b: Repo) => {
        const aValue = a.name as string
        const bValue = b.name as string
        return aValue.localeCompare(bValue)
      })
  }

  const searchRepos = async () => {
    setIsLoading(true)
    const results = await API.searchAll(searchValue.trim())

    if (results?.length) {
      const filteredResults = filterAndFormatResults(results)
      setReposListResults(filteredResults)
      setShowNoResultsMessage(filteredResults?.length === 0)
    } else {
      setShowNoResultsMessage(true)
    }
    setIsLoading(false)
  }

  // Search for repos based on user's search value after debounce timer
  useEffect(() => {
    if (searchValue.trim()?.length > 0) searchRepos() //Makes sure no to search an empty value
  }, [debouncedSearchValue])

  const handleSearch = async (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchValue(event.target.value)
  }

  const handleAdd = async (repo: Repo) => {
    //Storage limit validation
    if (reposList?.length >= 10) {
      setIsAlertVisible(true)
      setAlertInfo({
        message: 'Storage limit reached!',
        description:
          'Remove elements from the table to keep adding, you are allowed to add up to 10 items.',
        status: 'error',
      })
      //Duplicates validation
    } else if (reposList?.length && reposList.find((item: Repo) => item.id === repo.id)) {
      setIsAlertVisible(true)
      setAlertInfo({
        message: 'Duplicated Repository!',
        description: 'This repository already exist, try adding a diffetent one.',
        status: 'error',
      })
      // Saves selection to DB
    } else if (getIsReposerverRunning()) {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { name, description, ...restOfRepo } = repo
      const response = await actions.addRepo(restOfRepo)
      setIsAlertVisible(true)
      if (response?.id) {
        setReposListResults(reposListResults.filter((item: Repo) => item.id !== repo.id)) //Removes selected repo from dropdown list
        setAlertInfo({
          message: `${response?.fullName} was successfully added!`,
          description: '',
          status: 'success',
        })
      } else {
        setAlertInfo({
          message: `Failed adding repo to the list!`,
          description: '',
          status: 'error',
        })
      }
    }
  }

  const handleClearSearch = () => {
    setSearchValue('')
    setReposListResults([])
    setShowNoResultsMessage(false)
  }

  return isServerRunning ? (
    <>
      <AlertBar
        isOpen={isAlertVisible}
        message={alertInfo.message}
        description={alertInfo.description}
        status={alertInfo.status}
        onClose={() => setIsAlertVisible(false)}
      />
      <Container p={8}>
        <Box display="flex" alignItems="center">
          <SearchIcon w={5} h={5} color="gray.300" />
          <Input
            value={searchValue}
            onChange={handleSearch}
            px={6}
            mx={-5}
            variant="flushed"
            placeholder="Search your favorite repo here..."
          />
          {searchValue && (
            <CloseIcon
              onClick={handleClearSearch}
              color="gray.300"
              style={{
                cursor: 'pointer',
                position: 'relative',
              }}
            />
          )}
        </Box>
        {isLoading && (
          <Box display="flex" justifyContent="center" my={20}>
            <Spinner
              thickness="4px"
              speed="0.65s"
              emptyColor="gray.200"
              color="blue.500"
              size="xl"
            />
          </Box>
        )}
        {showNoResultsMessage && (
          <Box display="flex" justifyContent="center" my={20}>
            <Text fontSize="sm" color="orange.500">
              No results where found for: '{searchValue}'
            </Text>
          </Box>
        )}
        {reposListResults?.length > 0 && !isLoading && (
          <SearchRepoResults results={reposListResults} onSelect={handleAdd} />
        )}
      </Container>
    </>
  ) : (
    <Box display="flex" justifyContent="center" textAlign="center" p={8} my={20}>
      <Text fontSize="sm" color="orange.500">
        <WarningTwoIcon color="orange.500" mr={1} w={5} h={5} />
        {`Make sure to start REPOSERVER first.`} <br />
        {`Once started, reload the page to use the app!`}
      </Text>
    </Box>
  )
}

export default SearchRepo
