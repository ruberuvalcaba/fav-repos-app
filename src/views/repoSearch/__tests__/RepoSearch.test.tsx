import React from 'react'
import renderer from 'react-test-renderer'
import RepoSearch from '../RepoSearch'

it('RepoSearch component renders correctly', () => {
  const component = renderer.create(<RepoSearch />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
