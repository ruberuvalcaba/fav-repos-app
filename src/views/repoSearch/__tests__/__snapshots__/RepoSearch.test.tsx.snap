// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoSearch component renders correctly 1`] = `
[
  <style
    dangerouslySetInnerHTML={
      {
        "__html": ".css-wxprr5{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-pack:center;-ms-flex-pack:center;-webkit-justify-content:center;justify-content:center;text-align:center;padding:8px;margin-top:20px;margin-bottom:20px;}",
      }
    }
    data-emotion="css wxprr5"
  />,
  <div
    className="css-wxprr5"
  >
    <style
      dangerouslySetInnerHTML={
        {
          "__html": ".css-19qgjkk{font-size:sm;color:orange.500;}",
        }
      }
      data-emotion="css 19qgjkk"
    />
    <p
      className="chakra-text css-19qgjkk"
    >
      <style
        dangerouslySetInnerHTML={
          {
            "__html": ".css-1jwamw7{display:inline-block;line-height:1em;-webkit-flex-shrink:0;-ms-flex-negative:0;flex-shrink:0;color:orange.500;vertical-align:middle;margin-right:1px;width:5px;height:5px;}",
          }
        }
        data-emotion="css 1jwamw7"
      />
      <svg
        className="chakra-icon css-1jwamw7"
        focusable={false}
        viewBox="0 0 24 24"
      >
        <path
          d="M23.119,20,13.772,2.15h0a2,2,0,0,0-3.543,0L.881,20a2,2,0,0,0,1.772,2.928H21.347A2,2,0,0,0,23.119,20ZM11,8.423a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Zm1.05,11.51h-.028a1.528,1.528,0,0,1-1.522-1.47,1.476,1.476,0,0,1,1.448-1.53h.028A1.527,1.527,0,0,1,13.5,18.4,1.475,1.475,0,0,1,12.05,19.933Z"
          fill="currentColor"
        />
      </svg>
      Make sure to start REPOSERVER first.
       
      <br />
      Once started, reload the page to use the app!
    </p>
  </div>,
]
`;
